immutablecollections/_immutabledict.py

```diff
@@ -66,13 +69,46 @@ def immutabledict(
             f"Cannot create an immutabledict from {type(iterable)}, only {InstantiationTypes}"
         )
 
+    if forbid_duplicate_keys:
+        keys: List[KT]
+        # `for x in dict` grabs keys, but `for x in pairs` grabs pairs, so we must branch
+        if isinstance(iterable, Mapping):
+            # duplicate keys are possible if input is e.g. a multidict
```

Comment:
What types should we check for then? `dict_keys` has some `set` behavior but fails `isinstance`:
```
In [1]: d = {i:i*2 for i in range(10000)}

In [2]: type(d.keys())
Out[2]: dict_keys

In [3]: isinstance(d.keys(), set)
Out[3]: False

In [4]: from typing import Mapping

In [5]: isinstance(d, Mapping)
Out[5]: True

In [6]: from typing import Set

In [7]: isinstance(d.keys(), Set)
Out[7]: False
```

Reply:
We could branch on whether `iterable` has `.keys`, or we could have a weaker debug message that just says some unknown duplicate keys were present

Reply:
The code is actually fine; it's just the comment which is wrong :-)

Reply:
@qpwo : We do want to include the duplicate elements.  But we want to keep the "happy path" (which will be by far the most common) very fast; once we know there is an error, then we can do the work to figure out what the duplicate elements are (because no one cares if their crash takes an extra millisecond longer :-) )